rust/datafusion/src/physical_plan/string_expressions.rs

```diff
@@ -17,27 +17,100 @@
 
 //! String expressions
 
-use crate::error::{DataFusionError, Result};
-use arrow::array::{
-    Array, ArrayRef, GenericStringArray, StringArray, StringBuilder,
-    StringOffsetSizeTrait,
+use std::sync::Arc;
+
+use crate::{
+    error::{DataFusionError, Result},
+    scalar::ScalarValue,
+};
+use arrow::{
+    array::{Array, GenericStringArray, StringArray, StringOffsetSizeTrait},
+    datatypes::DataType,
 };
 
-macro_rules! downcast_vec {
-    ($ARGS:expr, $ARRAY_TYPE:ident) => {{
-        $ARGS
-            .iter()
-            .map(|e| match e.as_any().downcast_ref::<$ARRAY_TYPE>() {
-                Some(array) => Ok(array),
-                _ => Err(DataFusionError::Internal("failed to downcast".to_string())),
-            })
-    }};
+use super::ColumnarValue;
+
+pub(crate) fn unary_string_function<'a, T, O, F, R>(
```

Comment:
this function is effectively applying a `str->str` function `op` to all values in `args`? Some comments might be helpful